src/app/game-state/character.service.ts

```diff
@@ -234,6 +234,9 @@ export class CharacterService {
       'STANDARD', 'STORY');
     this.characterState.bloodlineCost *= 1000;
     this.characterState.bloodlineRank++;
+    if (this.characterState.bloodlineRank >= 4) {
```

Comment:
Hmm. The logic is here. Why do we also have it in the achievement?

Reply:
I found out that this function only being called when player upgrade their bloodline.

But what if a player originally already have bloodline 4 or 5, and the bloodlinecost already stored in their save file?

The achievement is trying to change the cost saved in their save file

Reply:
I not sure how the save file works, whether the save file save the bloodlinecost? or it save the bloodlineRank and reupgrade it back on loading? If it is saving the bloodlineRank and reupgrade when loading, then this function will make sure the bloodlineCost is correct

Reply:
my idea is, if the bloodline cost is saved in the save file, then when they load the save files, the new achievement will immediately pass the "check" function, so the effect will be applied to change the cost.

A very bad idea anyway, achievement should not be use like that

Reply:
Yeah. If you look in character.ts you can see the cost and the rank in the CharacterProperties. Anything in that interface gets saved into the save file.

Reply:
thats the reason why i add an achievement to make "correction" to the saved bloodline cost, because the old bloodline cost won't be affected by the change in the upgradeBloodline function

Reply:
Seems like we could just set the right cost here and be done. Whatever we set it to should be persisted in the save.

Reply:
The problem is, this function only runs when the player have enough spirit.

If the player don't have enough spirit, then they will be blocked at the store.service

So let's say a player done bloodline 4 before patch

So inside this function (before patch), the function will set bloodline 5 price to become 100T, and then lv up the bloodline rank from 3 to become 4.

The 100T is saved inside the save file.

So after we patch it, the player will still see 100T cost in the UI, when the try to click, they will also get a failed message from the store.service, and this function won't be get called.

-------------

Player who originally have bloodline  3 before we patch, will be normal. But player who originally have bloodline 4 & 5, will have the old 100T and 100q stored in their save file.

What my achievement try to do is to make sure that all those player who have bloodline 4 / 5 before patch, will get their value updated in the achievement effect.

Reply:
Ah, yeah, I see. Forgot about legacy save files. My bad.